From 4f47a2a72cd3e6803aaa115530571390e36cf24c Mon Sep 17 00:00:00 2001
From: bdeng3X <bingx.deng@intel.com>
Date: Fri, 12 Oct 2012 02:52:18 +0800
Subject: [PORT FROM MAIN] Re-connect NETLINK_KOBJECT_UEVENT socket when POLLERR happen in poll.

BZ: 58034

Bug description: USB driver hang after interval 0.2s quick plug/unplug stress test. re-connect to other PC or wall charger, DUT can't detect anything after stress test.
1. Print error log when POLLERR happen in poll.
2. Close socket when POLLERR occur. Then try to re-connect NETLINK_KOBJECT_UEVENT socket until it is success.

Change-Id: Iae5f98b91f6542f1abd45e2bde65384ef76d470e
Orig-Change-Id: I490fc833b3ac3c0b4f95c56074c11a8353a5a953
Signed-off-by: bdeng3X <bingx.deng@intel.com>
Reviewed-on: http://android.intel.com:8080/75940
Reviewed-by: lab_cactus <lab_cactus@intel.com>
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Reviewed-by: Li, XiaojingX <xiaojingx.li@intel.com>
Tested-by: Li, XiaojingX <xiaojingx.li@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 uevent/uevent.c |   18 ++++++++++++++++++
 1 files changed, 18 insertions(+), 0 deletions(-)

diff --git a/uevent/uevent.c b/uevent/uevent.c
index b9e3557..2eaa94a 100644
--- a/uevent/uevent.c
+++ b/uevent/uevent.c
@@ -25,7 +25,9 @@
 #include <sys/un.h>
 #include <sys/queue.h>
 #include <linux/netlink.h>
+#include "utils/Log.h"
 
+#define LOG_TAG "libhardware_legacy"
 
 LIST_HEAD(uevent_handler_head, uevent_handler) uevent_handler_list;
 pthread_mutex_t uevent_handler_list_lock = PTHREAD_MUTEX_INITIALIZER;
@@ -65,6 +67,14 @@ int uevent_init()
     return (fd > 0);
 }
 
+void uevent_exit()
+{
+    if (fd > 0) {
+        close(fd);
+        fd = -1;
+    }
+}
+
 int uevent_get_fd()
 {
     return fd;
@@ -92,6 +102,14 @@ int uevent_next_event(char* buffer, int buffer_length)
 
                 return count;
             } 
+        } else if ((fds.revents) & POLLERR) {
+            ALOGW("POLLERR: Error condition.");
+            uevent_exit();
+            int ret = 0;
+            do {
+                 ALOGI("Try to re-connect NETLINK_KOBJECT_UEVENT socket...");
+                 ret = uevent_init();
+            } while (ret != 1);
         }
     }
     
-- 
1.7.4.1

